refactor!: migrate `plugin-react-refresh` to `plugin-react`

aleclarson opened the pull request:
## ⚠️  Do not squash (to preserve commit history)

### Description

Once this plugin is released, `@vitejs/plugin-react-refresh` should be deprecated.

This plugin will do the following:
- Fast Refresh in dev only (same as in `plugin-react-refresh`)
- Support for [automatic JSX runtime](https://github.com/alloc/vite-react-jsx)
- Babel configuration that applies to dev *and* prod
  - Follow-up to #3319
  - Full configuration support (including presets, babelrc, etc)
- Warn if the old plugin or `vite-react-jsx` is also used

## Status

Almost ready to ship 🚢 

### Remaining questions

1. Should `jsxRuntime: "automatic"` be the default?
2. What should be done with `plugin-react/CHANGELOG.md`?

### Todo

- [x] Fix lint problems
- [x] Migrate to TypeScript
- [x] Copy over unit tests from [here](https://github.com/flying-sheep/babel-plugin-transform-react-createelement-to-jsx/blob/master/test/index.js?rgh-link-date=2021-06-11T23%3A12%3A58Z)
- [x] Setup a test suite in `./packages/playground` (in progress by @pengx17)

---

### What is the purpose of this pull request? <!-- (put an "X" next to an item) -->

- [ ] Bug fix
- [x] New Feature
- [ ] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vitejs/vite/blob/main/.github/contributing.md#pull-request-guidelines) and follow the [Commit Convention](https://github.com/vitejs/vite/blob/main/.github/commit-convention.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

underfin committed:
feat(plugin-react-refresh): support configure babel plugins
fix: #3276

aleclarson committed:
chore: rename plugin-react-refresh

This preserves the commit history.

aleclarson committed:
feat: integrate vite-react-jsx

..and improve Babel configuration.

aleclarson commented on packages/plugin-react/README.md:
````diff
@@ -25,6 +25,16 @@ export default {
+```js
+export default {
+  plugins: [reactRefresh({
+    plugins: ['@emotion/babel-plugin']
````
This example is outdated. Should be:

```js
reactPlugin({
  babel: {
    plugins: ['@emotion/babel-plugin']
  }
})
```

aleclarson commented on packages/plugin-react/src/fast-refresh.js:
```diff
@@ -0,0 +1,109 @@
+  prevRefreshReg = window.$RefreshReg$;
+  prevRefreshSig = window.$RefreshSig$;
+  window.$RefreshReg$ = (type, id) => {
+    RefreshRuntime.register(type, ${JSON.stringify(id)} + " " + id)
```
Oops :)

Shinigami92 commented:
I added now the new label `plugin: react`

Shinigami92 commented on packages/plugin-react/src/fast-refresh.js:
```diff
@@ -0,0 +1,109 @@
+  window.$RefreshSig$ = RefreshRuntime.createSignatureFunctionForTransform;
+}`.replace(/[\n]+/gm, '')
+
+export function addRefreshWrapper(code, accept) {
```
**suggestion:** We may need to disable `@typescript-eslint/explicit-module-boundary-types` for `js` files :thinking:

Shinigami92 replied:
Done via https://github.com/vitejs/vite/pull/3589

Please rebase after PR merged

aleclarson commented on packages/plugin-react/src/index.js:
```diff
@@ -0,0 +1,210 @@
+        let ast
+        if (id.endsWith('x')) {
+          if (opts.jsxRuntime === 'automatic') {
+            const [restoredAst, isCommonJS] = isNodeModules
```
todo: add a comment explaining what `restoreJSX` call is for

aleclarson commented on packages/plugin-react/src/index.js:
```diff
@@ -1,235 +1,210 @@
+          'classProperties',
+          'classPrivateProperties',
+          'classPrivateMethods',
+          'objectRestSpread'
```
Oops, this parser plugin can be removed, as it's enabled by default.

https://github.com/babel/babel/blob/813f2a3cce04a5366de7c1cc370652db64507305/packages/babel-plugin-syntax-typescript/src/index.js#L36-L38

aleclarson commented on packages/plugin-react/src/index.d.ts:
```diff
@@ -1,12 +1,29 @@
+   */
+  babel?: TransformOptions
+  /**
+   * @deprecated Use `babel.parserOpts.plugins` instead
```
Thoughts on deprecating `parserPlugins` in favor of `babel` option?

aleclarson commented on packages/plugin-react/src/index.d.ts:
```diff
@@ -1,12 +1,29 @@
+   * Enable `react-refresh` integration. Vite disables this in prod env or build mode.
+   * @default true
+   */
+  fastRefresh?: boolean
```
This option is for Rollup users. I made sure to keep this plugin compatible with Rollup, since it seemed to be designed that way in the first place.

aleclarson commented on packages/plugin-react/src/index.d.ts:
```diff
@@ -1,12 +1,29 @@
+
+export default viteReact
+
+export const preambleCode: string
```
Oops, I need to actually export this part. Is it exported for middleware mode, or what?

antfu commented on packages/plugin-react/package.json:
```diff
@@ -19,16 +28,18 @@
     "@babel/plugin-transform-react-jsx-source": "^7.12.13",
-    "react-refresh": "^0.9.0"
+    "react-refresh": "^0.9.0",
+    "vite-react-jsx": "^1.1.0"
```
Do you think we should inline it as the part of the package?

aleclarson replied:
Ideally, we should, but we need someone to either (a) migrate the new `plugin-react` to TypeScript or (b) migrate the `vite-react-jsx` code to JavaScript.

antfu replied:
Think we could migrate it into typescript if necessary.

Shinigami92 replied:
Migrate to TypeScript? I heard my name?! :eyes: 
What should I do? :laughing:

aleclarson replied:
Just need tsconfig.json, and have esbuild (in `build` script) use the ts loader, and add missing type annotations.

antfu replied:
We can take `plugin-vue` as a reference for TS setup I think. @aleclarson @Shinigami92 maybe you two can work on this PR together?

aleclarson replied:
and update the `exports` field in package.json

Shinigami92 replied:
I started with the rewrite here: https://github.com/aleclarson/vite/pull/1

aleclarson committed:
chore: link to vite in devDependencies

aleclarson committed:
chore: add missing @ts-check

aleclarson committed:
fix: runtime error

aleclarson committed:
fix: export preambleCode

aleclarson committed:
chore: remove superfluous parser plugin

aleclarson committed:
refactor: merge vite-react-jsx completely

Shinigami92 committed:
chore: rewrite index and fast-refresh

Shinigami92 committed:
chore: fix some errors

Shinigami92 committed:
chore: add tsconfig and add some todos

kjeske commented:
Thanks @aleclarson for your work in that PR, it's something I'm really looking forward to :) What is the status of the works?

aleclarson committed:
chore: fix typescript issues

aleclarson committed:
chore: add comment

aleclarson committed:
chore: fix remaining issues

aleclarson committed:
chore: tweak babel-restore-jsx

aleclarson committed:
chore: remove tsconfig paths

aleclarson committed:
build: use vite

aleclarson committed:
docs: update plugin-react readme

aleclarson commented:
@kjeske It should be ready for use. To get this merged, someone needs to setup some tests in `./packages/playground`. I don't know when I'll get around to doing that. I'm hoping a community member picks up the slack there, since I've done so much work already. :)

Shinigami92 commented:
`packages/plugin-react/CHANGELOG.md` may need to be resetted to v1.0.0 :thinking:

Shinigami92 committed:
chore: use newest deps versions

Shinigami92 committed:
chore: minor optimizations

Shinigami92 committed:
chore: use direct @babel/core types

Shinigami92 committed:
chore: get rid of any and improve typedefs

Shinigami92 committed:
chore: improve with generic

Shinigami92 committed:
chore: get rid of any

Shinigami92 committed:
chore: get rid of any

Shinigami92 commented on packages/plugin-react/src/jsx-runtime/babel-restore-jsx.ts:
```diff
@@ -0,0 +1,243 @@
+            getJSXIdentifier(prop.key)!,
+            getJSXAttributeValue(prop.value)
+          )
+        : t.jsxSpreadAttribute((prop as t.SpreadElement).argument)
```
This needs to be a `as t.SpreadElement` but could theoretically also be a `t.ObjectMethod`  
We should improve the flow analysis here or maybe fix it?!

Shinigami92 reviewed (commented):
`getJSXNode`, `getJSXName`, `getJSXProps` and `getJSXChildren` feels a bit to wide open :thinking: 
Maybe they can benefit from using generic arguments

aleclarson commented:
It might be better to copy over the unit tests from [here](https://github.com/flying-sheep/babel-plugin-transform-react-createelement-to-jsx/blob/master/test/index.js) *instead* of trying to perfect the types. I don't expect much future development of `babel-restore-jsx` to take place, so I don't think the types are worth the squeeze. I suppose I should've mentioned this before, but I didn't expect you to go the extra mile. :)

Shinigami92 commented:
Okay, so I'm glad that I just was bored yesterday :smile: 
But you should look at least into my first two review comments for `packages/plugin-react/src/fast-refresh.ts`

kjeske commented:
What's missing now in this PR to be completed?

aleclarson commented:
@kjeske See the todos in the OP. Feel free to open a PR targeting this one if you decide to help out 😃

pengx17 committed:
chore: update vite-dts

pengx17 committed:
build: use tsc instead of vite to build the plugin

pengx17 committed:
build: fix demo

pengx17 committed:
fix: automatic jsx runtime

pengx17 committed:
fix: esm/cjs build compatibility

pengx17 committed:
fix: fix review comments

pengx17 committed:
fix: a new way to check plugin conflicts

aleclarson committed:
fix: set `filter` variable immediately

...in case we are being used in Rollup but not Vite.

Commit:
chore: update lockfile

aleclarson committed:
Revert "chore: get rid of any and improve typedefs"

This reverts commit 57a1735ecf4dcc814bfc161593327cc170bf9ccb.

aleclarson committed:
Revert "chore: get rid of any"

This reverts commit 2c1b5cf92c0e34ba2e0f811ee9d4e906e4965204.

aleclarson committed:
Revert "chore: improve with generic"

This reverts commit 633b87a43740a38b8c9ade58a56308d60c78f34f.

aleclarson committed:
fix: flatten results of nested `getJSXProps` calls

aleclarson committed:
test: add tests for babel-restore-jsx

aleclarson committed:
chore: fix lint

Shinigami92 commented on packages/plugin-react/package.json:
```diff
@@ -19,17 +28,23 @@
+    "resolve": "^1.20.0"
+  },
+  "devDependencies": {
+    "vite": "link:../vite",
```
Should this be a peerDependency?

aleclarson replied:
I don't think it matters much?

Shinigami92 reviewed (commented):
There are some new lint warnings regarding to `===` can be `==` if it is compared with `null`
This is to check if it's `null` or `undefined` and is mostly kinda more safe, e.g. for refactorings

pengx17 committed:
feat(plugin-react): add a new playground to test jsx-runtime automic

pengx17 committed:
feat: import react also checks jsxInject

pengx17 committed:
build(plugin-react): add build-plugin-react to GitHub Actions

pengx17 committed:
chore: remove unused code

pengx17 committed:
fix: lint issues

pengx17 committed:
fix: add a warning when using jsxInject to import React

aleclarson committed:
chore: fix `dev` script

aleclarson committed:
chore: various tweaks

aleclarson committed:
chore: remove src/index.d.ts

aleclarson committed:
fix: use explicit extension with dynamic import

pengx17 committed:
build: also build-plugin-react at root's build

pengx17 committed:
build: use esbuild to bundle plugin-react

pengx17 committed:
fix: should not add legacy plugins to automatic runtime

pengx17 committed:
test: use emotion to demo react-jsx-runtime tests

pengx17 committed:
fix: move react/jsx-runtime optimization to vite:react-jsx

pengx17 committed:
fix: remove unnecessary react/jsx-runtime loader

aleclarson committed:
Revert "fix: remove unnecessary react/jsx-runtime loader"

This reverts commit f9006ba470ee20bd07cff76c5fdb01e56d3bccee.

aleclarson committed:
test: add react-switch to playground/react-jsx-runtime

This dependency is minified and commonjs, which will test the capabilities
of the `restoreJSX` function.

aleclarson committed:
fix: ensure node_modules is converted to automatic jsx-runtime

aleclarson committed:
fix: make vite:react-jsx work in Windows env

aleclarson committed:
test: fragment syntax

aleclarson committed:
refactor: simplify plugin imports

aleclarson committed:
refactor: consolidate `configResolved` hooks

Also:
- check `jsxInject` inside configResolved
- ensure all config values have reasonable defaults for Rollup compatibility

aleclarson committed:
fix: use development version of automatic jsx-runtime transform

aleclarson committed:
fix: ensure jsx source is added to elements in node_modules

pengx17 committed:
fix: tests

Sturgelose commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+          if (isReactModule && filter(id)) {
+            useFastRefresh = true
+            plugins.push([
+              await import('react-refresh/babel'),
```
```suggestion
              (await import('react-refresh/babel')).default,
```

Seems that this plugin does require to use the default export or it does not work.

Sturgelose commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+          }
+          // Even if the automatic JSX runtime is not used, we can still
+          // inject the React import for .jsx and .tsx modules.
+          else if (!isNodeModules && !/(^|\n)import React /.test(code)) {
+            code = `import React from 'react'; ` + code
+          }
```
When testing locally, having this enabled broke the build. The reason is that we already have the `import React from 'react'; ` in our code, so it duplicates the import and breaks stuff.

There should be an option to only do this if the user wants, as this kinda emulates the automatic JSX runtime, and this feature and changes are really recent for most of the projects to have it available.

Sturgelose commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        const result = ast
+          ? await babel.transformFromAstAsync(ast, code, babelOpts)
+          : await babel.transformAsync(code, babelOpts)
```
Instead of importing from babel, seems the compiler doesn't like it and prefers to import the function directly (otherwise complains that babel object doesn't have these functions/imports.

```suggestion
          ? await transformFromAstAsync(ast, code, babelOpts)
          : await transformAsync(code, babelOpts)
```

And in the top extend the babel import to 

```
import babel, { ParserOptions, TransformOptions, types as t, transformAsync, transformFromAstAsync } from '@b
abel/core'
```

Sturgelose reviewed (commented):
Hi, I did clone this branch locally to test if we could compile react + styled-components as it requires a babel plugin to make the css prop to work properly.

I was able to make it works and this plugin is just perfect, but I had to do a couple of changes to the code to make it work. Adding them in some comments, and I guess maybe we want some tests on them as well!

Also would be willing to help to have the PR ready as soon as possible as this would unlock us to use vite in our react projects :) 

Thanks a lot for the hard work!

pengx17 commented:
@Sturgelose hi, I have a pending [PR here](https://github.com/aleclarson/vite/pull/3) targeting to the current one. I hope it will cover all the issues you listed.

Sturgelose commented:
> @Sturgelose hi, I have a pending [PR here](https://github.com/aleclarson/vite/pull/3) targeting to the current one. I hope it will cover all the issues you listed.

Great, thanks a lot @pengx17 , didn't know on the work in this other PR. Following it then and resolving everything I spammed here!

iheyunfei commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+          }
+          // Even if the automatic JSX runtime is not used, we can still
+          // inject the React import for .jsx and .tsx modules.
+          else if (!isNodeModules && !/(^|\n)import React /.test(code)) {
```
Is this means that vite.config.json#esbuild.jsxInject will be deprecated?

iheyunfei commented on packages/plugin-react/src/index.d.ts:
```diff
@@ -0,0 +1,31 @@
+  fastRefresh?: boolean
+  /**
+   * Set this to `"automatic"` to use [vite-react-jsx](https://github.com/alloc/vite-react-jsx).
+   * @default "classic"
```
Would it be good to set default as "automatic" ? As a breaking change for new plugin release will be more acceptable.

iheyunfei commented on packages/plugin-react/src/fast-refresh.ts:
```diff
@@ -0,0 +1,109 @@
+export function isRefreshBoundary(ast: t.File): boolean {
+  // Every export must be a React component.
+  return ast.program.body.every((node) => {
+    if (node.type !== 'ExportNamedDeclaration') {
```
It looks like we ignore direct exports, no matter is anonymous or not. Should we need to follow the strategy mentioned in facebook/react#21181.

aleclarson replied:
I'm not sure. Maybe this can be addressed in a follow-up PR.

iheyunfei commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+    name: 'vite:babel',
+    enforce: 'pre',
+    async transform(code, id, ssr) {
+      if (/\.[tj]sx?$/.test(id)) {
```
I know there is a pattern people write JSX in `.js` but does it will happen on `.ts`?

iheyunfei replied:
Sorry. I found this is a vite react plugin not a vite react refresh plugin. It could be also used as a `babel-loader` .

iheyunfei commented on packages/plugin-react/src/index.ts:
```diff
@@ -0,0 +1,258 @@
+        if (!isNodeModules && !isProduction) {
+          plugins.push(
+            await interopDefault(
+              import('@babel/plugin-transform-react-jsx-self')
```
I didn't test it. But 'automatic' is not compatible with `@babel/plugin-transform-react-jsx-self` and `@babel/plugin-transform-react-jsx-source`. It's said in https://babeljs.io/docs/en/babel-preset-react .

aleclarson committed:
chore: fix dev build output dir

aleclarson committed:
chore: rename variables

aleclarson committed:
chore: rewrite option description

aleclarson commented:
> Is this means that vite.config.js#esbuild.jsxInject will be deprecated?

Probably not anytime soon

> Would it be good to set default as "automatic" ? As a breaking change for new plugin release will be more acceptable.

Good idea. WDYT @patak-js @antfu @yyx990803?

aleclarson committed:
docs: update readme

patak-js commented:
> Should jsxRuntime: "automatic" be the default?

+1. Even with the extra build time, if React is moving in this direction I think it is good that the default is automatic here, and it was said before in the thread let's use this opportunity to do the breaking changes.

> What should be done with plugin-react/CHANGELOG.md?

If there is a lot of shared history, we can copy the changelog from plugin-react-refresh and keep adding from there starting from the next minor version (like we did when we moved from create-app to create-vite). But probably to avoid confusion it would be better to start fresh and add an initial manual comment pointing to the previous plugin and changelog?

patak-js commented:
@aleclarson we discussed this PR and we are good to merge it 🎉 
About the remaining questions, if the automatic transform is stable in React then we can switch to it by default in the new plugin.
We can keep the old package there for a while, and do the redirect with a warning pointing to the new package. It is good that the changelog will be available a bit more. At one point, we can remove it from the repo.
For the new plugin, we can starts with v1.0 with a fresh changelog.

kjeske commented:
Since you are good to merge it, when could it actually happen? What is left to do?

patak-js commented:
What remains is to switch to automatic by default. And AFAIU @aleclarson wanted to do the last rebase before merging.

aleclarson committed:
feat: change `jsxRuntime` default to "automatic"

aleclarson committed:
chore: update changelog

aleclarson committed:
chore: rename `playground/react-jsx-runtime` to `react-emotion`

aleclarson committed:
chore: tweak `playground/react`

Shinigami92 commented on packages/plugin-react/README.md:
````diff
@@ -0,0 +1,96 @@
+
+```js
+// vite.config.js
+import react from '@vitejs/plugin-react'
+
+export default {
+  plugins: [react()]
+}
````
```suggestion
import react from '@vitejs/plugin-react'
import { defineConfig } from 'vite'

export default defineConfig({
  plugins: [react()]
})
```

aleclarson committed:
chore: use `defineConfig` in readme

Co-authored-by: Shinigami <chrissi92@hotmail.de>
